x-pack/filebeat/input/gcs/state.go

```diff
@@ -0,0 +1,94 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package gcs
+
+import (
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	maxFailedJobRetries int = 3
+)
+
+// state contains the the current state of the operation
+type state struct {
+	// Mutex lock to help in concurrent R/W
+	mu sync.Mutex
+	cp *Checkpoint
+}
+
+// Gcs sdks do not return results based on timestamps , but only based on lexicographic order
+// This forces us to maintain 2 different variables to calculate the exact checkpoint based on various scenarios
+type Checkpoint struct {
+	// name of the latest blob in alphabetical order
+	ObjectName string
+	// timestamp to denote which is the latest blob
+	LatestEntryTime *time.Time
+	// list of failed jobs due to unexpected errors/download errors
+	FailedJobs map[string]interface{}
+}
+
+func newState() *state {
+	return &state{
+		cp: &Checkpoint{
+			FailedJobs: make(map[string]interface{}),
+		},
+	}
+}
+
+// save, saves/updates the current state for cursor checkpoint
+func (s *state) save(name string, lastModifiedOn *time.Time) {
+	s.mu.Lock()
+	if s.cp.FailedJobs[name] == nil {
```

Comment:
If `FailedJobs` can be `map[string]int`
```suggestion
	if _, ok := s.cp.FailedJobs[name]; !ok {
```

Reply:
done